# Project 2: Boston Crime

Chris Bunker and Pat Steeves

Our goal with this project is to see if there is any correlation between sports and crime. We will use Boston, a city with a lot of recent sports success to see if times around championships or major losses/wins affect the crimes in certain areas in both quantity and type. 

## Problem 1: Examining the Data

In [1]:
!wget https://github.com/chrisbunker/istm-6212/blob/56f1ce54ba16316ee47f4e16bcbff0637cf4ad9c/boston_crime.zip?raw=true

--2016-11-11 16:19:34--  https://github.com/chrisbunker/istm-6212/blob/56f1ce54ba16316ee47f4e16bcbff0637cf4ad9c/boston_crime.zip?raw=true
Resolving github.com (github.com)... 192.30.253.112, 192.30.253.113
Connecting to github.com (github.com)|192.30.253.112|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://github.com/chrisbunker/istm-6212/raw/56f1ce54ba16316ee47f4e16bcbff0637cf4ad9c/boston_crime.zip [following]
--2016-11-11 16:19:34--  https://github.com/chrisbunker/istm-6212/raw/56f1ce54ba16316ee47f4e16bcbff0637cf4ad9c/boston_crime.zip
Reusing existing connection to github.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/chrisbunker/istm-6212/56f1ce54ba16316ee47f4e16bcbff0637cf4ad9c/boston_crime.zip [following]
--2016-11-11 16:19:34--  https://raw.githubusercontent.com/chrisbunker/istm-6212/56f1ce54ba16316ee47f4e16bcbff0637cf4ad9c/boston_crime.zip
Resolving raw.githubusercontent.com (raw.github

In [2]:
!unzip boston_crime.zip?raw=true

Archive:  boston_crime.zip?raw=true
  inflating: boston_crime.csv        


Now that our file on Boston Crimes from 7/8/2012 to 8/10/2015 and has been unzipped, lets see if it was done successfully so we can start to examine the csv file.

In [3]:
!csvcut -c2 boston_crime.csv | sort | uniq -c | sort -n | tail -10 

   6101 BE
   6406 IVPREM
   6432 IVMV
   8376 HITRUN
  11395 LARCEN
  12302 MVA
  14149 IVPER
  14204 IVPER 
  15932 ARREST
  15999 LARCRT


These are the most occuring natures of a crime that occurred in Boston from 2012 to 2015.

In [4]:
!csvcut -c3 boston_crime.csv | sort | uniq -c | sort -n | tail -10 

   8742 FRAUD
  12042 DRUG CHARGES
  12742 LARCENY FROM MOTOR VEHICLE
  12937 InvPer
  13339 VANDALISM
  13832 MVAcc
  17128 MedAssist
  17697 SIMPLE ASSAULT
  24443 OTHER LARCENY
  27363 VAL


These are the most occurring incident types that occurred in Boston from 2012 to 2015.

In [5]:
!csvcut -c2,3 boston_crime.csv | csvstat

  1. NatureCode
	<class 'str'>
	Nulls: True
	Unique values: 280
	5 most frequent values:
		IVPER:	28353
		LARCRT:	15999
		ARREST:	15932
		MVA:	15091
		LARCEN:	11395
	Max length: 6
  2. INCIDENT_TYPE_DESCRIPTION
	<class 'str'>
	Nulls: False
	Unique values: 127
	5 most frequent values:
		VAL:	27363
		OTHER LARCENY:	24443
		SIMPLE ASSAULT:	17697
		MedAssist:	17128
		MVAcc:	13832
	Max length: 32

Row count: 268056


### Column Definitions 

In [6]:
!csvcut -n boston_crime.csv

  1: COMPNOS
  2: NatureCode
  3: INCIDENT_TYPE_DESCRIPTION
  4: MAIN_CRIMECODE
  5: REPTDISTRICT
  6: REPORTINGAREA
  7: FROMDATE
  8: WEAPONTYPE
  9: Shooting
 10: DOMESTIC
 11: SHIFT
 12: Year
 13: Month
 14: DAY_WEEK
 15: UCRPART
 16: X
 17: Y
 18: STREETNAME
 19: XSTREETNAME
 20: Location


There are 20 different columns provided in the csv file for Boston Crimes. 
1. COMPNOS is used a primary key uniqiely identifying each of the records of an incident for easy reference. 
2. NatureCode is a reference code to the incident type
3. INCIDENT_TYPE_DESCRIPTION is the type of incident that occurred
4. MAIN_CRIMECODE is the code to the incident type
5. REPTDISTRICT is the district where the crime was reported
6. REPORTINGAREA is the area where the crime occurred (these are not area codes)
7. FROMDATE is the date of the reports
8. WEAPONTYPE is the type of weapon, if one was used
9. Shooting is a boolean function for if it was a shooting
10. DOMESTIC is a boolean function for if it was a domestic crime
11. SHIFT is the work shift in which the crime was reported
12. Year
13. Month
14. DAY_WEEK 
15. UCRPART is the type of crime that occurred (TYPE 1, 2, and 3) from highest severity to lowest
16. X ??
17. Y ?? 
18. STREETNAME
19. XSTREETNAME is the cross street
20. Location is the longitude and latitutde 

### Useful Columns

In [7]:
!csvcut -c1,2,3,4,5,6,7,8,9,10,15,18 boston_crime.csv | csvstat

  1. COMPNOS
	<class 'int'>
	Nulls: True
	Min: 110523814
	Max: 152052879
	Sum: 34718943844554
	Mean: 135728502.8540366
	Median: 130807210
	Standard Deviation: 9886105.346158931
	Unique values: 255796
	5 most frequent values:
		142052550:	2
		130614074:	1
		130021058:	1
		120473730:	1
		142022687:	1
  2. NatureCode
	<class 'str'>
	Nulls: True
	Unique values: 280
	5 most frequent values:
		IVPER:	28353
		LARCRT:	15999
		ARREST:	15932
		MVA:	15091
		LARCEN:	11395
	Max length: 6
  3. INCIDENT_TYPE_DESCRIPTION
	<class 'str'>
	Nulls: False
	Unique values: 127
	5 most frequent values:
		VAL:	27363
		OTHER LARCENY:	24443
		SIMPLE ASSAULT:	17697
		MedAssist:	17128
		MVAcc:	13832
	Max length: 32
  4. MAIN_CRIMECODE
	<class 'str'>
	Nulls: False
	Unique values: 77
	5 most frequent values:
		VAL:	27591
		06xx:	25731
		08xx:	18334
		MedAssist:	17837
		MVAcc:	15231
	Max length: 10
  5. REPTDISTRICT
	<class 'str'>
	Nulls: True
	Unique values: 13
	5 most frequent values:
		B2:	40640
		D4:	37908
		C11:	

From the above csvstat of the columns we find important we can see the class types, max values, etc. for for the table we will create in the next problem. We can also see which columns allow null values.

### Opportunities, Challenges, and Questions

From the data we have pulled from the csv file we believe we can successfully see if there is a correlation between sports and crime in Boston. We can cross reference the dates of sports games with dates of crimes and see which crimes have a higher chance of happening after sports wins vs. losses.

Some of the major challeneges we will have to deal with are:
   1. Data errors: some of the data seems to have been entered improperly with certain codes being placed in two rows. We may have to deal with this when we create out star schema.
   2. Incident names: we will need to find all of the incident codes and what they mean in order to fully understand the data.
   3. Making sure there is some proof of causation as well as correlation

## Problem 2: Defining and Exlporing a Database Model

Restart postgresql

In [8]:
!echo 'redspot' | sudo -S service postgresql restart

[sudo] password for jovyan: Restarting PostgreSQL 9.5 database server: main.


Create database

In [9]:
!createdb -U dbuser ex4

Connect to database

In [10]:
%load_ext sql

/opt/conda/lib/python3.5/site-packages/IPython/config.py:13: ShimWarning: The `IPython.config` package has been deprecated. You should import from traitlets.config instead.
  "You should import from traitlets.config instead.", ShimWarning)
/opt/conda/lib/python3.5/site-packages/IPython/utils/traitlets.py:5: UserWarning: IPython.utils.traitlets has moved to a top-level traitlets package.
  warn("IPython.utils.traitlets has moved to a top-level traitlets package.")


In [11]:
%sql postgresql://dbuser@localhost:5432/ex4

'Connected: dbuser@ex4'

Create first table. The datatypes are selected from the csvstat output in part 1

In [12]:
%%sql
DROP TABLE IF EXISTS crime;
CREATE TABLE crime (
    idnum INTEGER,
    nature_code VARCHAR(16),
    description VARCHAR(32),
    crime_code VARCHAR(32),
    district_code VARCHAR(4),
    report_area INTEGER,
    date_time TIMESTAMP,
    weapon VARCHAR(64),
    shooting BOOLEAN,
    domestic BOOLEAN,
    shift VARCHAR(6),
    year INTEGER,
    month INTEGER,
    day VARCHAR(10),
    severity VARCHAR(10),
    X NUMERIC,
    Y NUMERIC,
    street VARCHAR(64),
    cross_street VARCHAR(64),
    location VARCHAR(64)
);

Done.
Done.


[]

In [13]:
!pwd

/home/jovyan/work


In [14]:
%%sql
COPY crime FROM '/home/jovyan/work/boston_crime.csv'
CSV
HEADER
DELIMITER ',';

268056 rows affected.


[]

Ensure that all column types are correct and that all rows were loaded

In [15]:
%%sql
ALTER TABLE crime
DROP COLUMN idnum;

Done.


[]

In [16]:
%%sql
ALTER TABLE crime
ADD COLUMN idnum BIGSERIAL;

Done.


[]

In [17]:
%%sql
SELECT COUNT(*) FROM crime;

1 rows affected.


[(268056,)]

Correct capitalization error in column severity

In [18]:
%%sql 
SELECT DISTINCT severity FROM crime;

5 rows affected.


[('Part One',), ('Part three',), ('Part Three',), ('Other',), ('Part Two',)]

In [19]:
%%sql
UPDATE crime SET severity = 'Part Three' WHERE severity = 'Part three'

55482 rows affected.


[]

In [20]:
%%sql 
SELECT DISTINCT severity FROM crime;

4 rows affected.


[('Part One',), ('Part Three',), ('Other',), ('Part Two',)]

In [21]:
%%sql
SELECT year, COUNT(*) AS count FROM crime
GROUP BY year
ORDER BY count DESC;

4 rows affected.


[(2014, 88058), (2013, 87052), (2015, 49760), (2012, 43186)]

We can see that the number of crimes committed in 2013 and 2014 were similar. Since our data only spans about half of 2012 and half of 2015, the numbers appear to be substantially less in 2012 and 2015.

In [22]:
%%sql
SELECT month, COUNT(*) AS count FROM crime
GROUP BY month
ORDER BY count DESC;

12 rows affected.


[(7, 29113),
 (8, 25525),
 (5, 23294),
 (9, 23159),
 (10, 22836),
 (6, 22132),
 (3, 21295),
 (4, 20930),
 (1, 20697),
 (11, 20379),
 (12, 20235),
 (2, 18461)]

More crime seems to be committed during the warmer months of the year

## Problem 3: Defining a Star Schema

Please see Visio file for star schema diagram. This diagram displays the fact table and three dimensional tables that will be created below.

### Star Schema Tables

Create first version of fact table by simply copying granular data table

In [23]:
%%sql
DROP TABLE IF EXISTS crimes_fact;
CREATE TABLE crimes_fact AS
    (SELECT * FROM crime);

Done.
268056 rows affected.


[]

Create time dimension table using select attributes

In [24]:
%%sql
DROP TABLE IF EXISTS time_dimension;
CREATE TABLE time_dimension AS
    SELECT DISTINCT
    date_time,
    shift,
    year,
    month,
    day
FROM crime;

Done.
172958 rows affected.


[]

Since the time stamp is unique for all values, set it as primary key

In [25]:
%%sql
ALTER TABLE time_dimension ADD PRIMARY KEY (date_time);

Done.


[]

Create foreign key in facts table

In [26]:
%%sql
ALTER TABLE crimes_fact
ADD FOREIGN KEY (date_time) REFERENCES time_dimension(date_time);

Done.


[]

Create location dimension table using 3 select dimensions

In [27]:
%%sql
DROP TABLE IF EXISTS location_dimension;
CREATE TABLE location_dimension AS
    SELECT DISTINCT
    district_code,
    report_area,
    street
FROM crime;

Done.
20029 rows affected.


[]

Manually add primary key

In [28]:
%%sql
ALTER TABLE location_dimension ADD COLUMN location_key BIGSERIAL PRIMARY KEY;

Done.


[]

Create second version of fact table by selecting select columns of the joins of the old fact table and the location dimension table

In [29]:
%%sql
DROP TABLE IF EXISTS crimes_fact2;
CREATE TABLE crimes_fact2 AS
(
SELECT f.idnum, f.nature_code, f.description, f.crime_code, f.date_time, f.weapon, f.shooting, f.domestic, f.severity, d.location_key
FROM (SELECT * FROM location_dimension) AS d
INNER JOIN (SELECT * FROM crimes_fact) AS f
ON d.district_code = f.district_code
AND d.report_area = f.report_area
AND d.street = f.street
    )
;

Done.
268052 rows affected.


[]

Create crime type dimension table

In [30]:
%%sql
DROP TABLE IF EXISTS crime_dimension;
CREATE TABLE crime_dimension AS
    SELECT DISTINCT
    description,
    weapon,
    shooting,
    domestic,
    severity
FROM crime;

Done.
375 rows affected.


[]

Manually create primary key

In [31]:
%%sql
ALTER TABLE crime_dimension ADD COLUMN crime_key BIGSERIAL PRIMARY KEY;

Done.


[]

Using same logic as above, create the third version of the fact table

In [32]:
%%sql
DROP TABLE IF EXISTS crimes_fact3;
CREATE TABLE crimes_fact3 AS
(
SELECT f2.idnum, f2.location_key, f2.date_time, d.crime_key
FROM (SELECT * FROM crime_dimension) AS d
INNER JOIN (SELECT * FROM crimes_fact2) AS f2
ON d.description = f2.description
AND d.weapon = f2.weapon
AND d.shooting = f2.shooting
AND d.domestic = f2.domestic
AND d.severity = f2.severity
)
;

Done.
268052 rows affected.


[]

Take a look at the fact table

In [33]:
%%sql
SELECT * FROM crimes_fact3
LIMIT 30;

30 rows affected.


[(143201, 9679, datetime.datetime(2014, 2, 27, 19, 22), 1),
 (134437, 13318, datetime.datetime(2014, 1, 19, 1, 40), 2),
 (182427, 16854, datetime.datetime(2014, 8, 4, 12, 26), 2),
 (82300, 11740, datetime.datetime(2013, 6, 18, 0, 1), 2),
 (205704, 17318, datetime.datetime(2014, 11, 4, 17, 17), 2),
 (112958, 10137, datetime.datetime(2013, 10, 15, 16, 35), 2),
 (206235, 9364, datetime.datetime(2014, 11, 6, 22, 22), 2),
 (211714, 3814, datetime.datetime(2014, 12, 1, 9, 7), 2),
 (127201, 6856, datetime.datetime(2013, 12, 17, 8, 40), 2),
 (40256, 4395, datetime.datetime(2012, 12, 17, 11, 9), 2),
 (136303, 5547, datetime.datetime(2014, 1, 27, 11, 31), 2),
 (82917, 9296, datetime.datetime(2013, 6, 20, 12, 15), 2),
 (43383, 18460, datetime.datetime(2013, 1, 1, 9, 25), 2),
 (50378, 15747, datetime.datetime(2013, 2, 1, 11, 45), 2),
 (9490, 8930, datetime.datetime(2012, 8, 14, 8, 59), 2),
 (46073, 8291, datetime.datetime(2013, 1, 13, 14, 20), 2),
 (89910, 1173, datetime.datetime(2013, 7, 17, 10, 39), 2),
 (14911, 18628, datetime.datetime(2012, 9, 4, 4, 57), 2),
 (160016, 14255, datetime.datetime(2014, 5, 9, 16, 0), 2),
 (85386, 19046, datetime.datetime(2013, 6, 30, 0, 1), 2),
 (65538, 13942, datetime.datetime(2013, 4, 9, 21, 23), 2),
 (204187, 4927, datetime.datetime(2014, 10, 29, 14, 0), 2),
 (39028, 4530, datetime.datetime(2012, 12, 12, 14, 53), 2),
 (205821, 19199, datetime.datetime(2014, 11, 5, 7, 14), 2),
 (60476, 305, datetime.datetime(2013, 3, 19, 17, 26), 2),
 (93281, 3317, datetime.datetime(2013, 7, 30, 10, 9), 2),
 (47946, 8685, datetime.datetime(2013, 1, 21, 20, 52), 2),
 (158474, 473, datetime.datetime(2014, 5, 27, 9, 9), 2),
 (65452, 6555, datetime.datetime(2013, 4, 9, 13, 58), 2),
 (9748, 3317, datetime.datetime(2012, 8, 15, 8, 19), 2)]

Compare two of the above values with same dimension key (1786) to see if our logic works

In [34]:
%%sql
SELECT * FROM crimes_fact
WHERE idnum = 162349;

1 rows affected.


[('PARTY ', 'Gather', 'Gather', 'B2', 311, datetime.datetime(2014, 5, 18, 3, 13), 'Unarmed', False, False, 'Last', 2014, 5, 'Sunday', 'Part Three', Decimal('769791.8058'), Decimal('2940458.144'), 'SAVIN ST', None, '(42.31599931, -71.07988256)', 162349)]

In [35]:
%%sql
SELECT * FROM crimes_fact
WHERE idnum = 244996;

1 rows affected.


[('MVA', 'MVAcc', 'MVAcc', 'A1', 0, datetime.datetime(2015, 4, 30, 12, 50), 'Unarmed', False, False, 'Day', 2015, 4, 'Thursday', 'Other', Decimal('776522.8934'), Decimal('2958894.803'), 'SALEM ST', None, '(42.36649634, -71.05463956)', 244996)]

Both these crimes happened on Anderson Street In district A1 and report are 0!

Let's take a look at the fact table and all three dimension tables

In [36]:
%%sql
SELECT * FROM crimes_fact3
LIMIT 10;

10 rows affected.


[(143201, 9679, datetime.datetime(2014, 2, 27, 19, 22), 1),
 (134437, 13318, datetime.datetime(2014, 1, 19, 1, 40), 2),
 (182427, 16854, datetime.datetime(2014, 8, 4, 12, 26), 2),
 (82300, 11740, datetime.datetime(2013, 6, 18, 0, 1), 2),
 (205704, 17318, datetime.datetime(2014, 11, 4, 17, 17), 2),
 (112958, 10137, datetime.datetime(2013, 10, 15, 16, 35), 2),
 (206235, 9364, datetime.datetime(2014, 11, 6, 22, 22), 2),
 (211714, 3814, datetime.datetime(2014, 12, 1, 9, 7), 2),
 (127201, 6856, datetime.datetime(2013, 12, 17, 8, 40), 2),
 (40256, 4395, datetime.datetime(2012, 12, 17, 11, 9), 2)]

In [37]:
%%sql
SELECT * FROM time_dimension
LIMIT 10;

10 rows affected.


[(datetime.datetime(2014, 10, 13, 16, 48), 'First', 2014, 10, 'Monday'),
 (datetime.datetime(2012, 12, 9, 15, 0), 'Day', 2012, 12, 'Sunday'),
 (datetime.datetime(2013, 1, 21, 16, 4), 'First', 2013, 1, 'Monday'),
 (datetime.datetime(2013, 8, 9, 23, 0), 'First', 2013, 8, 'Friday'),
 (datetime.datetime(2012, 10, 4, 3, 50), 'Last', 2012, 10, 'Thursday'),
 (datetime.datetime(2012, 10, 25, 13, 0), 'Day', 2012, 10, 'Thursday'),
 (datetime.datetime(2013, 4, 9, 9, 55), 'Day', 2013, 4, 'Tuesday'),
 (datetime.datetime(2012, 11, 11, 22, 25), 'First', 2012, 11, 'Sunday'),
 (datetime.datetime(2013, 3, 21, 8, 50), 'Day', 2013, 3, 'Thursday'),
 (datetime.datetime(2014, 8, 31, 21, 3), 'First', 2014, 8, 'Sunday')]

In [38]:
%%sql
SELECT * FROM location_dimension
LIMIT 10;

10 rows affected.


[('A7', 34, 'JEFFRIES ST', 1),
 ('B2', 264, 'BLUE HILL AVE', 2),
 ('E18', 529, 'SPRAGUE ST', 3),
 ('C6', 206, 'B ST', 4),
 ('C11', 0, 'NAVILLUS TR', 5),
 ('C6', 0, 'BLACK FALCON AV', 6),
 ('B2', 331, 'GLENDALE ST', 7),
 ('B2', 285, 'PARKER ST', 8),
 ('A1', 67, 'GARDEN CT ST', 9),
 ('B2', 182, 'MALCOLM X BL', 10)]

In [39]:
%%sql
SELECT * FROM crime_dimension
LIMIT 10;

10 rows affected.


[('07RV', 'Knife', False, False, 'Other', 1),
 ('07RV', 'Other', False, False, 'Other', 2),
 ('07RV', 'Unarmed', False, False, 'Other', 3),
 ('32GUN', 'Firearm', False, False, 'Part Three', 4),
 ('32GUN', 'Knife', False, False, 'Part Three', 5),
 ('32GUN', 'Other', False, False, 'Part Three', 6),
 ('32GUN', 'Unarmed', False, False, 'Part Three', 7),
 ('AGGRAVATED ASSAULT', 'Firearm', False, False, 'Part One', 8),
 ('AGGRAVATED ASSAULT', 'Firearm', True, False, 'Part One', 9),
 ('AGGRAVATED ASSAULT', 'Knife', False, False, 'Part One', 10)]

## Problem 4: Exploring the Schema

Take a look at the most crime-ridden locations in Boston

In [40]:
%%sql
SELECT district_code, report_area, street
FROM location_dimension
WHERE location_key IN (
SELECT location_key
FROM crimes_fact3
GROUP BY location_key
ORDER BY COUNT(*) DESC
LIMIT 10
);

10 rows affected.


[('B3', 0, 'BLUE HILL AV'),
 ('C11', 0, 'DORCHESTER AV'),
 ('D4', 0, 'BOYLSTON ST'),
 ('B2', 282, 'WASHINGTON ST'),
 ('D4', 0, 'MASSACHUSETTS AV'),
 ('B2', 0, 'WASHINGTON ST'),
 ('A1', 0, 'WASHINGTON ST'),
 ('E18', 0, 'HYDE PARK AV'),
 ('D14', 0, 'COMMONWEALTH AV'),
 ('D4', 0, 'HARRISON AV')]

What about the hours of the day where the most crime happens?

In [41]:
%%sql
SELECT hours, COUNT(*) FROM (
    SELECT EXTRACT(HOUR FROM date_time) AS hours
    FROM crimes_fact3
    ) AS hour
GROUP BY hours
ORDER BY count DESC
;

24 rows affected.


[(0.0, 20316),
 (12.0, 17390),
 (17.0, 16084),
 (18.0, 15956),
 (16.0, 15835),
 (14.0, 13500),
 (19.0, 13486),
 (9.0, 13006),
 (15.0, 13000),
 (11.0, 12870),
 (10.0, 12827),
 (13.0, 12636),
 (20.0, 12370),
 (8.0, 11581),
 (21.0, 10963),
 (22.0, 10406),
 (23.0, 8541),
 (1.0, 8229),
 (7.0, 8091),
 (2.0, 7280),
 (3.0, 4196),
 (6.0, 4126),
 (4.0, 2767),
 (5.0, 2596)]

Or which days of the week sees the most crime?

In [42]:
%%sql
SELECT day, COUNT(*) FROM (
    SELECT d.day
    FROM time_dimension AS d, crimes_fact3 AS f
    WHERE d.date_time = f.date_time
    ) AS days
GROUP BY day
ORDER BY count DESC
;

7 rows affected.


[('Friday', 42031),
 ('Thursday', 38627),
 ('Saturday', 38446),
 ('Wednesday', 38391),
 ('Monday', 38255),
 ('Tuesday', 37910),
 ('Sunday', 34392)]

Now let's take a look and see whether there is more crime after Boston sports teams win championships

### Red Sox 2013 World Series Title

First we will compare the number of crimes after the world series win compared to the number of crimes of the 6 days that are before the start of the world series.

In [43]:
%%sql
SELECT count(idnum) AS "Post WS Crimes"
FROM crimes_fact3, time_dimension
WHERE time_dimension.date_time BETWEEN '10-29-2013 20:00:00' AND '11-4-2013 20:00:00'
AND crimes_fact3.date_time = time_dimension.date_time;

1 rows affected.


[(1551,)]

In [56]:
%%sql
SELECT count(idnum) AS "Random Week Crimes"
FROM crimes_fact3, time_dimension
WHERE crimes_fact3.date_time = time_dimension.date_time
AND time_dimension.date_time BETWEEN '10-17-2013 20:00:00' AND '10-23-2013 20:00:00';

1 rows affected.


[(1417,)]

Lets also look at the crimes DURING the world series.

In [45]:
%%sql
SELECT count(idnum) AS "World Series Crimes"
FROM crimes_fact3, time_dimension
WHERE time_dimension.date_time BETWEEN '10-23-2013 20:00:00' AND '10-29-2013 20:00:00'
AND crimes_fact3.date_time = time_dimension.date_time;

1 rows affected.


[(1346,)]

We can now see that during the World Series there were more than 200 less crimes than after the World Series. The week after the World Series actually had more crime than the week prior to the start of the world series!

In order for us to be able to get the average of severity of the crime for certain days and weeks, we need to create a new table that has the severity strings and matches them to the integer of 0,1,2, or 3 to correspond to the correct level. 0 will refer to other as it will not have much use in the average of severity because we do not know what it is. We need to create a new table first with the distinct idnum and severity, then create a new colunm and update it to have integers corresponding to the 4 levels. We do this below:

In [46]:
%%sql
DROP TABLE IF EXISTS crime_severity;
CREATE TABLE crime_severity AS
    SELECT DISTINCT
    idnum,
    severity
FROM crime;

Done.
268056 rows affected.


[]

In [47]:
%%sql
ALTER TABLE crime_severity ADD COLUMN severity_num INTEGER;

Done.


[]

In [48]:
%%sql
SELECT DISTINCT severity, severity_num FROM crime_severity;

4 rows affected.


[('Other', None), ('Part One', None), ('Part Two', None), ('Part Three', None)]

In [49]:
%%sql
UPDATE crime_severity SET severity_num = 1 WHERE severity = 'Part One'

65261 rows affected.


[]

In [50]:
%%sql
UPDATE crime_severity SET severity_num = 2 WHERE severity = 'Part Two'

98341 rows affected.


[]

In [51]:
%%sql
UPDATE crime_severity SET severity_num = 3 WHERE severity = 'Part Three'

89005 rows affected.


[]

In [52]:
%%sql
UPDATE crime_severity SET severity_num = 0 WHERE severity = 'Other'

15449 rows affected.


[]

In [53]:
%%sql
SELECT DISTINCT severity, severity_num FROM crime_severity;

4 rows affected.


[('Part Two', 2), ('Part Three', 3), ('Other', 0), ('Part One', 1)]

Now we can use this table to take averages of the severity levels.

In [59]:
%%sql
SELECT AVG (crime_severity.severity_num) AS "Post World Series Crime Severity"
FROM crime_severity, crime_dimension, time_dimension, crimes_fact3
WHERE time_dimension.date_time BETWEEN '10-29-2013 20:00:00' AND '11-5-2013 20:00:00'
AND crimes_fact3.date_time = time_dimension.date_time
AND crimes_fact3.crime_key = crime_dimension.crime_key
AND crime_severity.severity = crime_dimension.severity;

1 rows affected.


[(Decimal('2.0885504804122496'),)]

In [60]:
%%sql
SELECT AVG (crime_severity.severity_num) AS "World Series Crime Severity"
FROM crime_severity, crime_dimension, time_dimension, crimes_fact3
WHERE time_dimension.date_time BETWEEN '10-23-2013 20:00:00' AND '10-29-2013 20:00:00'
AND crimes_fact3.date_time = time_dimension.date_time
AND crimes_fact3.crime_key = crime_dimension.crime_key
AND crime_severity.severity = crime_dimension.severity;

1 rows affected.


[(Decimal('2.0289480079932053'),)]

We can see that the average severity of crime the week after winning the world series was about a 2, but the average severity of during the world series is less by .06. Now lets see for all other times. Because the sample size is too large though, the notebook cannot allocate enough memory to calculate it, so we will take the week before the worlds series started to compare. If too many people are using datanotebook this will still not be able to run, but without the limit it takes too long anyways. As you will see, it has almost the same severity average as during the world series.

In [57]:
%%sql
SELECT AVG (crime_severity.severity_num) AS "Average Crime Severity"
FROM crime_severity, crime_dimension, time_dimension, crimes_fact3
WHERE time_dimension.date_time BETWEEN '10-17-2013 20:00:00' AND '10-23-2013 20:00:00'
AND crimes_fact3.date_time = time_dimension.date_time
AND crimes_fact3.crime_key = crime_dimension.crime_key
AND crime_severity.severity = crime_dimension.severity;

1 rows affected.


[(Decimal('2.0216879780484057'),)]

### 2014 Patriots Super Bowl Win

First we will do the same as we did with the World Series and take the counts for crime the week after the super bowl and 7 random days not including those after the Super Bowl. We will also look at crimes the day of the super bowl.

In [61]:
%%sql
SELECT count(idnum) AS "Post Super Bowl Crimes"
FROM crimes_fact3, time_dimension
WHERE time_dimension.date_time BETWEEN '02-01-2015 20:00:00' AND '02-07-2015 20:00:00'
AND crimes_fact3.date_time = time_dimension.date_time;

1 rows affected.


[(1360,)]

In [62]:
%%sql
SELECT count(idnum) AS "Pre Super BowlCrimes"
FROM crimes_fact3, time_dimension
WHERE time_dimension.date_time BETWEEN '01-26-2013 20:00:00' AND '02-01-2013 20:00:00'
AND crimes_fact3.date_time = time_dimension.date_time;

1 rows affected.


[(1473,)]

In [63]:
%%sql
SELECT count(idnum) AS "Super Bowl Crimes"
FROM crimes_fact3, time_dimension
WHERE time_dimension.date_time BETWEEN '02-01-2015 00:00:00' AND '02-02-2015 00:00:00'
AND crimes_fact3.date_time = time_dimension.date_time;

1 rows affected.


[(220,)]

1From above we can see that the week after the Super Bowl actually had more than 100 less crimes than the week after the world series win. Now this could also be due to weather factors but it could also show the super bowls affect on peoples behaviors.

Next we will again look at the severity of the crimes the week after the super bowl 

In [64]:
%%sql
SELECT AVG (crime_severity.severity_num) AS "Super Bowl Crime Severity"
FROM crime_severity, crime_dimension, time_dimension, crimes_fact3
WHERE time_dimension.date_time BETWEEN '02-01-2015 20:00:00' AND '02-07-2015 20:00:00'
AND crimes_fact3.date_time = time_dimension.date_time
AND crimes_fact3.crime_key = crime_dimension.crime_key
AND crime_severity.severity = crime_dimension.severity;

1 rows affected.


[(Decimal('2.1983568637343942'),)]

In [65]:
%%sql
SELECT AVG (crime_severity.severity_num) AS "Pre Super Bowl Crime Severity"
FROM crime_severity, crime_dimension, time_dimension, crimes_fact3
WHERE time_dimension.date_time BETWEEN '01-26-2013 20:00:00' AND '02-01-2013 20:00:00'
AND crimes_fact3.date_time = time_dimension.date_time
AND crimes_fact3.crime_key = crime_dimension.crime_key
AND crime_severity.severity = crime_dimension.severity;

1 rows affected.


[(Decimal('2.2272423349078175'),)]

An interesting thing to see is that even though there were less crimes committed after the super bowl than the world series, the severity of those crimes were quite a bit higher almost 2.2 compared to 2.0. The pre Super Bowl severity is .03 higher than post super bowl crime severity.

## Bonus: Augmenting Dimensions

### Weather's Affect on Crime

Weather can have a major impact on crimes committed. We could see that just by the amount of crimes committed in each month; the summer months were much higher than the winter months. 